zanata-war/src/main/java/org/zanata/security/SecurityFunctions.java

```diff
@@ -66,29 +64,34 @@
  *         href="mailto:camunoz@redhat.com">camunoz@redhat.com</a>
  */
 @Slf4j
-public class SecurityFunctions {
-    protected SecurityFunctions() {
-    }
+public class SecurityFunctions extends PermissionProvider {
+
+    @Inject
+    private ZanataIdentity identity;
+
+    @Inject
+    @Authenticated
+    private HAccount authenticatedAccount;
 
     /* admin can do anything */
     @GrantsPermission
-    public static boolean isAdmin() {
-        return getIdentity().hasRole("admin");
+    public boolean isAdmin() {
+        return identity.hasRole("admin");
     }
 
-    public static boolean isProjectMaintainer(HProject project) {
+    public boolean isProjectMaintainer(HProject project) {
         return isProjectRole(project, Maintainer);
     }
 
-    public static boolean isProjectTranslationMaintainer(HProject project) {
+    public boolean isProjectTranslationMaintainer(HProject project) {
         return isProjectRole(project, TranslationMaintainer);
     }
 
     /* Check whether the authenticated person has the given role in the given
      * project.
      */
-    private static boolean isProjectRole(HProject project, ProjectRole role) {
-        Optional<HAccount> account = getAuthenticatedAccount();
+    private boolean isProjectRole(HProject project, ProjectRole role) {
+        Optional<HAccount> account = Optional.fromNullable(authenticatedAccount);
```

Comment:
We would have to do that (the `@PostContstruct`) everywhere which injects
the authenticated account and wants to use Optional. I've been seeing this
sort of thing all over the place with authenticated account, so it would be
nice to limit the ugliness to the producer.

Besides, that producer doesn't look ugly to me. Just a little unfamiliar,
because we're still getting used to working with CDI producers.



On 15 February 2016 at 12:07, Carlos Munoz <notifications@github.com> wrote:

> In zanata-war/src/main/java/org/zanata/security/SecurityFunctions.java
> <https://github.com/zanata/zanata-server/pull/1093#discussion_r52854273>:
>
> >          return isProjectRole(project, TranslationMaintainer);
> >      }
> >
> >      /* Check whether the authenticated person has the given role in the given
> >       * project.
> >       */
> > -    private static boolean isProjectRole(HProject project, ProjectRole role) {
> > -        Optional<HAccount> account = getAuthenticatedAccount();
> > +    private boolean isProjectRole(HProject project, ProjectRole role) {
> > +        Optional<HAccount> account = Optional.fromNullable(authenticatedAccount);
>
> @seanf <https://github.com/seanf> After talking this over with Patrick, I
> think I like better the idea of a @PostConstruct which builds the Optional
> object. The resulting Producer would be rather nasty (it injects the same
> qualifier as it is producing, and the generic type of the optional is lost):
>
> @Produces @AuthenticatedOptional<HAccount> getAuthenticatedAccount(@Authenticated HAccount account) {
>    return Optional.ofNullable(account);
> }
>
> all this just to create a simple nullable.
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/1093/files#r52854273>.
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat